### PR TITLE
LUCENE-10464, LUCENE-10477: WeightedSpanTermExtractor.extractWeightedSpanTerms to rewrite sufficiently

### PR DESCRIPTION
**Pull request by cpoerschke**
* https://issues.apache.org/jira/browse/LUCENE-10464: don't unnecessarily rewrite the same query in a loop
* https://issues.apache.org/jira/browse/LUCENE-10477: rewrite 'recursively' via a loop

### CODE REVIEW

**Comment by jpountz** [comment]
This change makes sense to me.

File: lucene/highlighter/src/java/org/apache/lucene/search/highlight/WeightedSpanTermExtractor.java [hunk 1]
```diff
@@ -309,11 +309,11 @@ protected void extractWeightedSpanTerms(
     final IndexSearcher searcher = new IndexSearcher(getLeafContext());
     searcher.setQueryCache(null);
     if (mustRewriteQuery) {
+      final SpanQuery rewrittenQuery = (SpanQuery) spanQuery.rewrite(getLeafContext().reader());
```
Comment (jpountz) [inline lucene/highlighter/src/java/org/apache/lucene/search/highlight/WeightedSpanTermExtractor.java:312]: Based on the discussion we had on LUCENE-10477, let's rewrite in a for-loop like Indexsearcher#rewrite does? https://github.com/apache/lucene-solr/blob/releases/lucene-solr/8.0.0/lucene/core/src/java/org/apache/lucene/search/IndexSearcher.java#L665-L671

File: lucene/highlighter/src/java/org/apache/lucene/search/highlight/WeightedSpanTermExtractor.java [hunk 2]
```diff
@@ -309,11 +309,12 @@ protected void extractWeightedSpanTerms(
     final IndexSearcher searcher = new IndexSearcher(getLeafContext());
     searcher.setQueryCache(null);
     if (mustRewriteQuery) {
+      final SpanQuery rewrittenQuery =
+          (SpanQuery) IndexSearcher.rewrite(spanQuery, getLeafContext().reader());
```
[the remark this answers is not in the document]
Comment (cpoerschke) [reply lucene/highlighter/src/java/org/apache/lucene/search/highlight/WeightedSpanTermExtractor.java:313]: Actually `searcher` from line 309 would be suitable then too actually?!

**Review by jpountz — APPROVED** [review]

File: lucene/highlighter/src/java/org/apache/lucene/search/highlight/WeightedSpanTermExtractor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jpountz) [reply lucene/highlighter/src/java/org/apache/lucene/search/highlight/WeightedSpanTermExtractor.java]: +1

**Pull request merged**